File: 141_linked_list_cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    if (!head || !head.next){
        return false;
    }

    let one = two = head;

    while (one && one.next){
        one = one.next.next;
        two = two.next;

        if (one === two){
            return true;
        }
    }
    return false;
};
//Time Complexity O(n), Space Complexity O(1);

File: 141_linked_list_cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return False
        
        one = two = head
        while one is not None and one.next is not None:
            one = one.next.next
            two = two.next

            if one == two:
                return True
        return False